File: typescript/src/grpcClient.ts
import * as grpc from "@grpc/grpc-js";
import * as protoLoader from "@grpc/proto-loader";
import { performance } from "perf_hooks";
import { Empty } from "./proto/dtos_pb";
import path from "path";

const testCounts = [100, 200, 300];

const PROTO_PATH = path.join(__dirname, "proto/dtos.proto");
const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
  keepCase: true,
  longs: String,
  enums: String,
  defaults: true,
  oneofs: true,
});

const protoDescriptor = grpc.loadPackageDefinition(packageDefinition) as any;
const musicaAppProto = protoDescriptor.musica_app;

async function testLoad(
  client: any,
  method: string,
  args: any,
  name: string,
  numTest: number
) {
  const start = performance.now();

  for (let i = 0; i < numTest; i++) {
    await new Promise<void>((resolve, reject) => {
      client[method](args, (error: grpc.ServiceError, response: any) => {
        if (error) {
          console.error(`Error: ${error.message}`);
          reject(error);
        } else {
          resolve(response);
        }
      });
    });
  }

  const end = performance.now();
  console.log(`${name} : ${(end - start) / 1000} seconds`);
}

export async function runTests() {
  const clientUsuario = new musicaAppProto.UsuarioService(
    "localhost:5003",
    grpc.credentials.createInsecure()
  );
  const clientMusica = new musicaAppProto.MusicaService(
    "localhost:5003",
    grpc.credentials.createInsecure()
  );
  const clientPlaylist = new musicaAppProto.PlaylistService(
    "localhost:5003",
    grpc.credentials.createInsecure()
  );

  for (let numTest of testCounts) {
    console.log(`${numTest} requests:`);

    await testLoad(
      clientUsuario,
      "LerUsuarios",
      new Empty(),
      "Usuarios",
      numTest
    );

    await testLoad(clientMusica, "LerMusicas", new Empty(), "Musicas", numTest);

    await testLoad(
      clientPlaylist,
      "LerPlaylists",
      { usuario_id: 1 },
      "Playlists",
      numTest
    );
  }
}

File: rust/src/grpc_client.rs
use std::time::Instant;
use tonic::{transport::Channel, Request};

pub mod musica_app {
    tonic::include_proto!("musica_app");
}

use musica_app::musica_service_client::MusicaServiceClient;
use musica_app::playlist_service_client::PlaylistServiceClient;
use musica_app::usuario_service_client::UsuarioServiceClient;
use musica_app::Empty;

pub async fn run_tests() -> Result<(), Box<dyn std::error::Error>> {
    let test_counts = vec![100, 200, 300];
    let channel = Channel::from_static("http://localhost:5003")
        .connect()
        .await?;

    let mut usuario_client = UsuarioServiceClient::new(channel.clone());
    let mut musica_client = MusicaServiceClient::new(channel.clone());
    let mut playlist_client = PlaylistServiceClient::new(channel.clone());

    for &count in &test_counts {
        println!("Running tests with {} requests:", count);

        let start = Instant::now();
        for _ in 0..count {
            let request = Request::new(Empty {});
            let res = usuario_client.ler_usuarios(request).await;
            if let Err(err) = res {
                eprintln!("Request failed: {:?}", err);
            }
        }
        let duration = start.elapsed();
        println!(
            "Usuarios load test with {} requests completed in {:.2?} seconds",
            count, duration
        );

        let start = Instant::now();
        for _ in 0..count {
            let request = Request::new(Empty {});
            let res = musica_client.ler_musicas(request).await;
            if let Err(err) = res {
                eprintln!("Request failed: {:?}", err);
            }
        }
        let duration = start.elapsed();
        println!(
            "Musicas load test with {} requests completed in {:.2?} seconds",
            count, duration
        );

        let start = Instant::now();
        for _ in 0..count {
            let request = Request::new(Empty {});
            let res = playlist_client.ler_playlists(request).await;
            if let Err(err) = res {
                eprintln!("Request failed: {:?}", err);
            }
        }
        let duration = start.elapsed();
        println!(
            "Playlists load test with {} requests completed in {:.2?} seconds",
            count, duration
        );
    }

    Ok(())
}
